internal/rcm/api.go

```diff
@@ -93,7 +96,7 @@ func (api *API) submit(writer http.ResponseWriter, request *http.Request, _ http
 
 	// JSON structure expected from the client
 	var composeRequest struct {
-		Distribution  common.Distribution   `json:"distribution"`
+		Distribution  string                `json:"distribution"`
```

Comment:
> `ComposeRequest` included a `common.Distribution`, which had to be
resolved in PushComposeRequest. Use a real `distro.Distro` object here,
and push resolving it to the rcm package. That makes more sense, because
the input is validated closer to where it is received now. It also
simplifies `PushComposeRequest`.

This is only partially correct. The fact that you replaced `common.Distribution` here with `string` actually moved the validation further away, because one point of having the common types is to do their validation immediately at user's input and then only operate with them instead of strings.

If you still want to get rid of `common.Distribution` isn't it possible to write this line like this:
```
		Distribution  distro.Distro                `json:"distribution"`
```
That is write a common unmarshaler for `distro.Distro`?

Reply:
I referred to (captial) `ComposeRequest` in that commit message, not (lower-case) `composeRequest`. Sorry for the confusion. I've reworded the commit message to explain this change better.

> [...] because one point of having the common types is to do their validation immediately at user's input and then only operate with them instead of strings.

Right, but because we get the real `Distro` object in the same function now, I ended up with code like this:

```go
name, ok := composeRequest.Distribution.ToString()
// error handling

distro := api.distros.GetDistro(name)
// error handling
```

This effectively validates distro twice, and needlessly converts a `string` to an `int`, just to convert it back to a `string`. It doesn not buy any type-safety either, because go doesn't have enums.

> That is write a common unmarshaler for `distro.Distro`?

That's not possible, because unmarshallers cannot have access to surrounding state (we'd need a `distro.DistroRegistry`.

Reply:
If we cannot unmarshal to `distro.Distro` directly, I'd prefer to instead change the `api.distros.GetDistro()` to accept `common.Distribution` instead of a string, it is ready for the change. It just wasn't a priority yet.

So the code will become:
```
distro := api.distros.GetDistro(composeRequest.Distribution)
// error handling
```
And we simply cannot get rid of the error handling here.

Reply:
I don't think that is a good idea. It would mean that
1. We validate twice: once the `string` that's coming in from the request, turning it into an int. And the again turning the int in `GetDistro` to a `Distro`. This seems redundant and complex.
2. The function signature would become `GetDistro(distro common.Distribution) Distro`. I think the current one—`GetDistro(name string) Distro`—is much simpler to understand.

Reply:
But look at the `GetDistro` implementation. I agree that the signature looks weird, but in that case we should change the name, for example:
```
ToDistroInterface(distributionTag common.Distribution) Distro
```
it's not an awesome example, but represents the idea.